RLS support

brson opened the pull request:
I think this is sufficient for basic support.

With this PR, rustup will create 'rls' proxies in CARGO_HOME/bin, and will do so lazily when that binary doesn't exist.

After this PR one can write

```
rustup component add rls
rustup component add rust-src
rustup component add rust-analysis
```

to get a working RLS, assuming the 'rls' package is deployed.

Next steps are to make `rustup component` accept multiple components, so RLS can be installed with just `rustup compnent add rls rust-src rust-analysis` (cc @durka).

I'd suggest that RLS have nice error handling for cases where rust-src and rust-analysis aren't available. It might suggest running the proper rustup commands.

rustup probably needs to emit a better error message when somebody tries to run `rls` without installing it. Since the proxy always exists, it will say something like "the binary 'rls' doesn't exist in toolchain X". It would be better for rustup to know which package contains 'rls' and say so.

cc @nrc @jonathandturner

brson committed:
Add the rls bin to installation

brson committed:
Fix the self-update tests to use the correct version of the bins

I realized that, even in the installer tests, most of the binaries
being run are out of the clitest 'exedir', not from CARGO_HOME/bin.
So this patch fixes that by modifying the generic test runner
to run bins off of the PATH, with PATH configured by default prefixed
with 'exedir'. The cli-self-upd tests extend PATH with CARGO_HOME/bin.

brson commented:
Hm, this has a bug in it from the changes to the test suite.

brson committed:
Fix test suite bug

Diggsey commented:
Is `add_rls_proxy` strictly necessary? Won't the new proxy be automatically created anyway when rustup updates? (it always calls `install_bins`)

brson commented:
> Is add_rls_proxy strictly necessary? Won't the new proxy be automatically created anyway when rustup updates (it always calls install_bins)

The self-update will not on its own create the new proxy. `install_bins` is run on upgrade but it is the _old_ install bins that does not know about rls.

Diggsey commented:
Isn't it the new `install_bins` that gets run? The previous rustup calls the new one with `--self-replace`, and that one updates the hardlinks.

brson commented:
Uh, yeah I think you are right about that. Wow. I should sleep on this.

brson committed:
Roll back test suite changes, roll back lazy rls upgrade

The lazy upgrade isn't needed, per Diggsey. The test suite changes
are too complex to deal with right now, and introduce the risk
of the environment interfering with the tests (which is why the
tests are all red on the bots but green locally).

brson commented:
OK, I pushed some hasty changes to try to fix the failing tests, and remove the lazy upgrade.

Attempting to make the test suite run bins off the PATH ended up causing all kinds of problems with the environment seeping in, so I just rolled that back. rustup is really frustrating to test...

brson commented:
@Diggsey r?

I've really lost perspective on this patch, just throwing crap at it and hoping it works out.

Diggsey commented:
@bors r+ LGTM

bors commented:
:pushpin: Commit 1c9f146 has been approved by `Diggsey`

<!-- @bors r=Diggsey 1c9f146ee0ed877c65ef4451a6f2be40bc0661a0 -->

bors commented:
:hourglass: Testing commit 1c9f146 with merge 4e0c2a4...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/brson/rustup-rs/build/1.0.916)

brson commented:
@bors retry

weird error:

```
---- multirust_upgrade_works_with_proxy stdout ----
	thread 'multirust_upgrade_works_with_proxy' panicked at 'assertion failed: fs::symlink_metadata(&multirust_dir).unwrap().file_type().is_symlink()', tests\cli-rustup.rs:739
```

bors commented:
:hourglass: Testing commit 1c9f146 with merge 55ea19a...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/brson/rustup-rs/build/1.0.920), [status-travis](https://travis-ci.org/rust-lang-nursery/rustup.rs/builds/215777942)
Approved by: Diggsey
Pushing 55ea19ad6a8845a60fa3854c6e1e89886a6becf9 to master...
